### PR TITLE
Handle refusal event

### PR DESCRIPTION
**Pull request by AdeHarrison**
## Motivation and Context
Receipt message structure not matching event spec

## What has changed
Receipt message structure change as per event spec

## How to test?
[as per](https://github.com/ONSdigital/census-rm-case-processor/pull/30)

## Links
[Trello](https://trello.com/c/29ZsK7ci/825-rmc-197-handle-refusal-received-event-rev13-8)

### CODE REVIEW

**Comment by AdeHarrison** [comment]
[RMC-197 Handle Refusal Received Event [REV13] (8)](https://trello.com/c/29ZsK7ci/825-rmc-197-handle-refusal-received-event-rev13-8)

**Comment by AdeHarrison** [comment]
[RMC-197 Handle Refusal Received Event [REV13] (8)](https://trello.com/c/29ZsK7ci/825-rmc-197-handle-refusal-received-event-rev13-8)

**Review by jcox-dev — DISMISSED** [review]
`camelCase` variable names should be converted to `snake_case`

File: app/subscriber.py [hunk 1]
```diff
@@ -43,7 +41,7 @@ def receipt_to_case(message: Message):
     try:
         payload = json.loads(message.data)  # parse metadata as JSON payload
         metadata = payload['metadata']
-        case_id, tx_id = metadata['case_id'], metadata['tx_id']
+        case_id, tx_id, questionnaire_id = metadata['case_id'], metadata['tx_id'], metadata['questionnaire_id']
```
Comment (davidmort) [inline app/subscriber.py]: this line will error if case_id (not a mandatory field) isn't in the payload, this was fixed in this pr: https://github.com/ONSdigital/census-rm-pubsub/pull/28

**Review by davidmort — CHANGES_REQUESTED** [review]

File: app/subscriber.py [hunk 2]
```diff
@@ -58,7 +56,28 @@ def receipt_to_case(message: Message):
...
+    responseObject['caseId'] = case_id
+    responseObject['questionnaireId'] = questionnaire_id
+    responseObject['unreceipt'] = 'false'
+
+    payloadObject['response'] = responseObject
+
+    receiptMessage['event'] = eventObject
+    receiptMessage['payload'] = payloadObject
+
+    send_message_to_rabbitmq(json.dumps(receiptMessage))  # NB: in the future this should hold `questionnaire_id`
```
Comment (davidmort) [inline app/subscriber.py]: remove the comment on this line as its not applicable anymore

**Review by davidmort — CHANGES_REQUESTED** [review]

File: app/subscriber.py (hunk 2 above)
Comment (AdeHarrison) [reply app/subscriber.py]: done

File: app/subscriber.py (hunk 1 above)
Comment (AdeHarrison) [reply app/subscriber.py]: done

File: app/subscriber.py [hunk 3]
```diff
@@ -58,7 +58,28 @@ def receipt_to_case(message: Message):
     log = log.bind(questionnaire_id=questionnaire_id, created=time_obj_created, tx_id=tx_id, case_id=case_id)
 
     metadata['response_datetime'] = time_obj_created
-    send_message_to_rabbitmq(json.dumps(metadata))  # NB: in the future this should hold `questionnaire_id`
+
+    receiptMessage = {}
+    eventObject = {}
+    payloadObject = {}
+    responseObject = {}
```
Comment (AdamHawtin) [inline app/subscriber.py]: These dictionaries could be replaced by a literal definition which lets it read more like the JSON it's destined to become:
```python
    receipt_message = {
        'event': {
            'type': 'RESPONSE_RECEIVED',
            'source': 'RECEIPT_SERVICE',
            'channel': 'EQ',
            'dateTime': time_obj_created,
            'transactionId': tx_id
        },
        'payload': {
            'response': {
                'caseId': case_id,
                'questionnaireId': questionnaire_id,
                'unreceipt': False
            }
        }
    }
```

File: app/subscriber.py [hunk 4]
```diff
@@ -58,7 +58,28 @@ def receipt_to_case(message: Message):
...
+
+    eventObject['type'] = 'RESPONSE_RECEIVED'
+    eventObject['source'] = 'RECEIPT_SERVICE'
+    eventObject['channel'] = 'EQ'
+    eventObject['dateTime'] = time_obj_created
+    eventObject['transactionId'] = tx_id
+
+    responseObject['caseId'] = case_id
+    responseObject['questionnaireId'] = questionnaire_id
+    responseObject['unreceipt'] = 'false'
```
Comment (AdamHawtin) [inline app/subscriber.py]: Should be a `False` boolean object and let the json module serialize it?

**Review by AdamHawtin — DISMISSED** [review]

File: app/subscriber.py [hunk 5]
```diff
@@ -58,7 +58,28 @@ def receipt_to_case(message: Message):
     log = log.bind(questionnaire_id=questionnaire_id, created=time_obj_created, tx_id=tx_id, case_id=case_id)
 
     metadata['response_datetime'] = time_obj_created
```
Comment (AdamHawtin) [inline app/subscriber.py]: There's no reason to modify the metadata dictionary here now, it doesn't get used again.

File: app/subscriber.py (hunk 3 above)
Comment (AdeHarrison) [reply app/subscriber.py]: done

File: app/subscriber.py (hunk 4 above)
Comment (AdeHarrison) [reply app/subscriber.py]: done

File: app/subscriber.py (hunk 5 above)
Comment (AdeHarrison) [reply app/subscriber.py]: done

**Review by jcox-dev — APPROVED** [review]

**Review by AdamHawtin — APPROVED** [review]

**Review by davidmort — APPROVED** [review]

**Pull request merged**